### PR TITLE
add ability to customize HTTP error response for unary RPCs

### PR DESCRIPTION
**Pull request by jhump**
The provides a way for calling code to set a custom error renderer, to override the default mapping of gRPC errors to HTTP responses.

Since streams write the gRPC status into the body, as a trailing message, this is only for unary RPCs.

It also tweaks the existing mapping so that a `Cancelled` or `DeadlineExceeded` code only maps to 499 when the request context is cancelled. Otherwise, a 502 Bad Gateway or 504 Gateway Timeout status is used, respectively.

### CODE REVIEW

File: channel.go [hunk 1]
```diff
@@ -1,22 +1,14 @@
...
 // transport, or an in-process transport.
-type Channel interface {
-	// InvokeUnary executes a unary RPC, sending the given req message and populating
-	// the given resp with the server's reply.
-	Invoke(ctx context.Context, methodName string, req, resp interface{}, opts ...grpc.CallOption) error
-
-	// InvokeStream executes a streaming RPC.
-	NewStream(ctx context.Context, desc *grpc.StreamDesc, methodName string, opts ...grpc.CallOption) (grpc.ClientStream, error)
-}
+type Channel = grpc.ClientConnInterface
```
Comment (jhump) [inline channel.go:11]: Maybe this should be a separate PR?

But now that gRPC has its own version of this interface, using a type alias may make it easier to use grpchan in other contexts...

File: httpgrpc/server.go [hunk 2]
```diff
@@ -181,23 +270,23 @@ func HandleStream(svr interface{}, serviceName string, desc *grpc.StreamDesc, st
...
 			tr.Details = statProto.Details
-		} else if err != nil {
-			tr.Code = int32(codes.Internal)
-			tr.Message = "Internal Server Error"
-		}
-
-		// must put something the trailing message so it's size is non-zero
-		// (otherwise, it's envelope cannot indicate a negative number, and
-		// it will be confused for a normal response message)
-		if tr.Message == "" {
```
Comment (jhump) [inline httpgrpc/server.go:199]: I've reformulated the error handling above to match the unary error handling. And I structured it in a way that the `tr` message can never be empty: if there was an error, it will have a non-zero code to serialize at a minimum; if there was no error, it will have an "OK" message to serialize.

File: httpgrpc/server.go [hunk 3]
```diff
@@ -101,7 +178,11 @@ func HandleMethod(svr interface{}, serviceName string, desc *grpc.MethodDesc, un
 		if err != nil {
 			st, _ := status.FromError(err)
 			if st.Code() == codes.OK {
-				st = status.New(codes.Internal, err.Error())
```
Comment (jhump) [inline httpgrpc/server.go:104]: This old formulation was losing any details. So now we preserve them and basically just convert to a new status proto where only the code is changed.

File: channel.go (hunk 1 above)
Comment (dragonsinth) [reply channel.go:11]: SGTM assuming they're identical; otherwise it could break implementors.
Should we deprecate this symbol however, in favor fo the grpc type?

File: httpgrpc/server.go [hunk 4]
```diff
@@ -34,29 +34,106 @@ import (
...
+// given function to render an error.  It is only used for unary RPCs since
+// streaming RPCs serialize a status message to the response trailer (in the
+// HTTP body) instead.
+//
+// The function can examine and add response header values via the supplied
+// http.Header. It can also examine the request context. It must then return
+// the HTTP status code to use.
+//
+// If no such option is used the handler will use DefaultErrorRenderer.
+func ErrorRenderer(errFunc func(context.Context, *status.Status, http.Header) (httpCode int)) HandlerOption {
```
Comment (dragonsinth) [inline httpgrpc/server.go:55]:
```suggestion
func ErrorRenderer(errFunc func(reqCtx context.Context, st *status.Status, rspHdr http.Header) (httpCode int)) HandlerOption {
```

**Review by dragonsinth — APPROVED** [review]

File: channel.go (hunk 1 above)
Comment (jhump) [reply channel.go]: Good call. I've pulled it out for now and will put it another branch, along with deprecation.

**Comment by jhump** [comment]
@dragonsinth, I pushed another commit because I realized that we should treat "Cancelled" similarly to "DeadlineExceeded".

**Comment by dragonsinth** [comment]
Not as certain about Cancelled.. I can see an argument for always 499'ing that.

**Comment by jhump** [comment]
> Not as certain about Cancelled.. I can see an argument for always 499'ing that.

If the client has _not_ actually cancelled, I don't think 499 is ever appropriate. That code is really for internal book-keeping as opposed to a code that we send to the client. (The client shouldn't ever be able to see one, since it means "client has closed the request".)
